modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java

```diff
@@ -0,0 +1,157 @@
+        if (i >= 0) {
+            SqlIdentifier id = (SqlIdentifier) columnNames.get(i);
+            throw validator.newValidationError(id,
+                    RESOURCE.aliasListDuplicate(id.getSimple()));
+        }
+
+        TableDescriptor descriptor = table.unwrap(TableDescriptor.class);
+
+        assert descriptor != null;
+
+        RelDataType rowTypeSansHidden = descriptor.rowTypeSansHidden();
+        // At the moment, only system columns are hidden, and user must provide
+        // aliases for all non-system columns.
+        if (columnNames.size() != rowTypeSansHidden.getFieldList().size()) {
+            // Position error over all column names
+            SqlNode node = operands.size() == 3
+                    ? operands.get(2)
+                    : new SqlNodeList(columnNames, SqlParserPos.sum(columnNames));
+
+            throw validator.newValidationError(node,
+                    RESOURCE.aliasListDegree(rowTypeSansHidden.getFieldCount(),
+                            getString(rowTypeSansHidden), columnNames.size()));
+        }
+
+        // Although this may look like proper way to address `system` columns, in fact it's not:
+        //     1) Calcite has different notion of `system` columns. First of all, all implementations
+        //     of namespace just propagate call to getRowType() method. Second, system columns are
+        //     registered only when converter meets LogicalJoin.
+        //
+        //     2) Row type returned from this method will be used to resolve columns. Thus, if we skip
+        //     `system` columns validator won't be able to resolve any reference of such columns (given
+        //     that they are registered only in case of joins)
+        RelDataType rowType = childNs.getRowTypeSansSystemColumns();
+
+        assert rowType != null;
+
+        int fieldsCount = rowType.getFieldCount();
+        Mapping mapping = Mappings.create(MappingType.PARTIAL_FUNCTION, fieldsCount, nameList.size());
+        int idx = 0;
+        for (ColumnDescriptor column : descriptor) {
+            if (!column.hidden()) {
+                mapping.set(column.logicalIndex(), idx++);
+            }
+        }
+
+        RelDataType aliasedType = validator.getTypeFactory().builder()
+                .addAll(
+                        Util.transform(
+                                rowType.getFieldList(),
+                                f -> {
+                                    int index = mapping.getTargetOpt(f.getIndex());
+                                    String name = index == -1 ? f.getName() : nameList.get(index);
+
+                                    return Pair.of(name, f.getType());
+                                }
+                        )
+                )
+                .kind(rowType.getStructKind())
+                .build();
+
+        return validator.getTypeFactory()
+                .createTypeWithNullability(aliasedType, rowType.isNullable());
```

Comment:
I understand that this is a copy-paste from the calcite, but just curious what is the difference between
```
return createTypeWithNullability(aliasedType, rowType.isNullable());
```
and
```
return aliasedTypeBuilder...nullable(rowType.isNullable()).build();
```

Reply:
```
    /**
     * Changes the nullability of the last field added.
     *
     * @throws java.lang.IndexOutOfBoundsException if no fields have been
     *                                             added
     */
    public Builder nullable(boolean nullable) {
      RelDataType lastType = types.get(types.size() - 1);
      if (lastType.isNullable() != nullable) {
        final RelDataType type =
            typeFactory.createTypeWithNullability(lastType, nullable);
        types.set(types.size() - 1, type);
      }
      return this;
    }
```

it changes nullability for the last column, but we need to make an entire row nullable

Reply:
:woman_facepalming:  thanks